### ISSUE TITLE
`sym()` resolves symbols in aggregations, but not in printf

### ISSUE THREAD

**Issue by Birch-san** [iovisor/bpftrace#209]
I'm on commit dac74162bce7a941a90cf97b85b341c636c0530c, Ubuntu 18.04 LTS.

I tried to resolve kernel symbols. Aggregations seem to behave differently than do printf()s.

### Aggregation (successfully prints symbol)

```bash
# terminal A
sudo bpftrace -e 'tracepoint:skb:kfree_skb {
  @[sym(args->location)] = count()
}
interval:s:1 {
  print(@);
  clear(@)
}'

# terminal B
curl -s 'http://bbc.co.uk'

# triggers the following output from terminal A:
Attaching 2 probes...
@[unix_stream_connect]: 2
```

### printf() (prints empty string)

```bash
# terminal A
sudo bpftrace -e 'tracepoint:skb:kfree_skb {
  printf("symbol: %s\n", sym(args->location))
}'

# terminal B
curl -s 'http://bbc.co.uk'

# triggers the following output from terminal A:
Attaching 1 probe...
symbol: 
```

For reference, this is what the guide says about symbol translation:

> Some of these are asynchronous: the kernel queues the event, but some time later (milliseconds) it is processed in user-space. The asynchronous actions are: printf(), time(), and join(). Both sym() and usym(), as well as the variables stack and ustack, record addresses synchronously, but then do symbol translation asynchronously.

**Comment by brendangregg** [comment]
Thanks for the issue. It seems that other builtins have broken as well, eg, username. I think the issue is in bpftrace.cpp, in BPFtrace::get_arg_values(), with these code sections:

```
        resolved_symbols.emplace_back(strdup(
              resolve_sym(*(uint64_t*)(arg_data+arg.offset)).c_str()));
        arg_values.push_back((uint64_t)resolved_symbols.back().get());
```

But I don't know the fix yet, or what broke it. These used to work...

**Comment by brendangregg** [comment]
If I avoid the resolved_symbols.emplace_back() and just do this:

```
arg_values.push_back((uint64_t)strdup(resolve_sym(*(uint64_t*)(arg_data+arg.offset)).c_str()));
```

it works. I don't know why the emplace_back()/back().get() stopped working (@ajor ?), plus I'm not sure we're even using resolved_symbols yet (looks like it was meant to be a cache, but it is incomplete).

**Comment by Birch-san** [comment]
> looks like it was meant to be a cache, but it is incomplete

Here's the original commit which introduced `resolved_symbols`:

https://github.com/iovisor/bpftrace/commit/0e82af015b22e4c7d7d33c8815102285b182d558#diff-d231d684666d47519cd093494342e228R52

It's just to keep in scope the `std::string` returned by `resolve_sym()`, so that the memory that `.c_str()` points to doesn't get deallocated before we print `arg_values`.

**Comment by Birch-san** [comment]
Oh, I just noticed that (as of a month ago) we `return arg_values`:

https://github.com/iovisor/bpftrace/commit/7d490e547281e6c7a4600abfc814b01c6260dc42#diff-d231d684666d47519cd093494342e228R390

That's surely the problem. `arg_values` holds raw pointers to our copied strings. Our copied strings will get deallocated when we return from the function (final reference to unique_ptr is lost).

**Comment by gila** [comment]
I was trying to use `usym` today, and it suffers from the same bug I assume. `printf(....)` did not work when doing trying to use `usym(arg0)` (the first argument is a callback function). The workaround by using aggregations did work, however. Added the comment just in case and I was using commit 0a7dcadc8861e9eaa5483fa202bd6155a902dd14 on Ubuntu 18.04.

**Comment by Birch-san** [comment]
@gila it's fixed in https://github.com/iovisor/bpftrace/pull/301; awaiting review.

**Comment by gila** [comment]
@Birch-san Thanks. I've checked out your code and checked it out:

before:

```
# bpftrace -e 'uprobe:./a.out:call { printf("addr: %p\n name: %s\n addr: %p\n", arg0, usym(arg0), uaddr("cb"));}'
Attaching 1 probe...
addr: 0x55659f7d269a
 name: 0��V
 addr: 0x69a
^C
```

after:

```
# bpftrace -e 'uprobe:./a.out:call { printf("addr: %p\n name: %s\n addr: %p\n", arg0, usym(arg0), uaddr("cb"));}'
Attaching 1 probe...
addr: 0x55d3c78fb69a
 name: 0x55d3c78fb69a
 addr: 0x69a
^C
```

**Comment by danobi** [comment]
Merged #301
